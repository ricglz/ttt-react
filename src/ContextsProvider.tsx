import { useTranslation } from "react-i18next";
import { Toaster } from "react-hot-toast";
import { BrowserRouter as Router } from "react-router-dom";
import type { FC, ReactNode } from "react";

type Locale = string;

function getDirection(locale: Locale) {
  return locale === "ar" ? ["rtl", "text-right"] : ["ltr", "text-left"];
}

type Props = {
  ChildrenComponent: FC;
  children?: ReactNode;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  RouterComponent?: any;
};

export default function ContextsProvider({
  children,
  ChildrenComponent,
  RouterComponent = Router,
}: Props) {
  const { i18n } = useTranslation();
  const [direction, klass] = getDirection(i18n.language);

  return (
    <div dir={direction} className={klass}>
      <RouterComponent>
        <ChildrenComponent />
        {children}
      </RouterComponent>
      <Toaster position="bottom-right" />
    </div>
  );
}
