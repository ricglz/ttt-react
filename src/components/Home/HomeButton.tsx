import type { Locale } from 'react-i18next';
import { Link } from 'react-router-dom';
import { FormattedMessage } from '../Layout/FormattedText';

type RenderTextProps =
  | { staticText?: false; text: Locale }
  | { staticText?: true; text: string };

function renderText({ staticText, text }: RenderTextProps) {
  return staticText ? (
    <span>{text}</span>
  ) : (
    <FormattedMessage locale={text as Locale} />
  );
}

type Props = { url: string } & RenderTextProps;

const HomeButton = ({ url, ...rest }: Props) => (
  <div className="row justify-content-center border-top py-3">
    <Link className="btn btn-home" to={url}>
      {renderText(rest)}
    </Link>
  </div>
);

export default HomeButton;
