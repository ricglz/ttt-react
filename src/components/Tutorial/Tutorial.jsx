import React from 'react';
import PropTypes from 'prop-types';
import exImage1 from '../../images/example1.gif';
import exImage2 from '../../images/example2.png';
import BackButton from '../Layout/BackButton';
import { FormattedParagraph, FormattedHeader2 } from '../Layout/FormattedText';

const Image = ({ src }) => (
  <div className="col">
    <img src={src} className="img-fluid" alt="example" />
  </div>
);

Image.propTypes = {
  src: PropTypes.string.isRequired,
};

function renderParagraphs() {
  const paragraphsTitles = [
    'tutorial.first',
    'tutorial.second',
    'tutorial.third',
  ];
  return paragraphsTitles.map((title) => (
    <FormattedParagraph key={title} locale={title} />
  ));
}

const Tutorial = () => (
  <div className="container">
    <div className="row text-center pb-3 border-bottom">
      <div className="col">
        <FormattedHeader2 locale="shared.tutorial" />
      </div>
    </div>
    <div className="pt-3" />
    {renderParagraphs()}
    <div className="row text-center">
      <Image src={exImage1} />
    </div>
    <FormattedParagraph locale="tutorial.fourth" />
    <div className="row text-center">
      <Image src={exImage2} />
    </div>
    <FormattedParagraph locale="tutorial.fifth" />
    <div className="row text-center">
      <BackButton text="shared.back" url="/" />
    </div>
  </div>
);

export default Tutorial;
