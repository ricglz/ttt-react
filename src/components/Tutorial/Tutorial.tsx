import React from 'react';
import exImage1 from '../../images/example1.gif';
import exImage2 from '../../images/example2.png';
import BackButton from '../Layout/BackButton';
import { FormattedParagraph, FormattedHeader2 } from '../Layout/FormattedText';

type ImageProps = {
  src: string
};

const Image = ({ src }: ImageProps) => (
  <div className="col">
    <img src={src} className="img-fluid" alt="example" />
  </div>
);

const PARAGRAPHS_TITLES = [
  'tutorial.first',
  'tutorial.second',
  'tutorial.third',
];

const Tutorial = () => (
  <div className="container">
    <div className="row text-center pb-3 border-bottom">
      <div className="col">
        <FormattedHeader2 locale="shared.tutorial" />
      </div>
    </div>
    <div className="pt-3" />
    {
      PARAGRAPHS_TITLES.map((title) => (
        <FormattedParagraph key={title} locale={title} />
      ))
    }
    <div className="row text-center">
      <Image src={exImage1} />
    </div>
    <FormattedParagraph locale="tutorial.fourth" />
    <div className="row text-center">
      <Image src={exImage2} />
    </div>
    <FormattedParagraph locale="tutorial.fifth" />
    <div className="row text-center">
      <BackButton text="shared.back" url="/" />
    </div>
  </div>
);

export default Tutorial;
