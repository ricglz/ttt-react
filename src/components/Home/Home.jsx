import React from 'react';
import { FormattedMessage } from 'react-intl';
import HomeButton from './HomeButton';

function Home() {
  return (
    <div className="container text-center">
      <div className="row">
        <div className="col">
          <FormattedMessage id="homePage.title" default="Home Page">
            {txt => <h1>{txt}</h1>}
          </FormattedMessage>
        </div>
      </div>
      <HomeButton text="shared.sp" url="/singleplayer" />
      <HomeButton text="shared.mp" url="/multiplayer" />
      <HomeButton staticText text="Online" url="/online" />
      <HomeButton text="shared.tutorial" url="/tutorial" />
      <HomeButton staticText text="Contributors" url="/contributors" />
    </div>
  );
}

export default Home;
