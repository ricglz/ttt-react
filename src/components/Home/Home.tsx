import React from 'react';
import { FormattedHeader } from '../Layout/FormattedText';
import HomeButton from './HomeButton';

function Home() {
  return (
    <div className="container text-center">
      <div className="row">
        <div className="col">
          <FormattedHeader locale="homePage.title" defaultMessage="Home Page" />
        </div>
      </div>
      <HomeButton text="shared.sp" url="/singleplayer" />
      <HomeButton text="shared.mp" url="/multiplayer" />
      <HomeButton staticText text="Online" url="/online" />
      <HomeButton text="shared.tutorial" url="/tutorial" />
      <HomeButton staticText text="Contributors" url="/contributors" />
    </div>
  );
}

export default Home;
