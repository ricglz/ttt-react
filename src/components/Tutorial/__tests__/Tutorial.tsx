import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import Home from '../../Home/Home';
import Tutorial from '../Tutorial';
import ContextsProvider from '../../../ContextsProvider';

function Component() {
  return (
    <ContextsProvider
      ChildrenComponent={Home}
      RouterComponent={MemoryRouter}
    >
      <Tutorial />
    </ContextsProvider>
  );
}

it('renders without crashing', () => {
  const component = <Component />;
  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot();
});
