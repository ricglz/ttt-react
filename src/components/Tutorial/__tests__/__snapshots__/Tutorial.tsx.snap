// Vitest Snapshot v1

exports[`renders without crashing 1`] = `
<div
  className="text-left"
  dir="ltr"
>
  <div
    className="container text-center"
  >
    <div
      className="row"
    >
      <div
        className="col"
      >
        <h1>
          Home Page
        </h1>
      </div>
    </div>
    <div
      className="row justify-content-center border-top py-3"
    >
      <a
        className="btn btn-home"
        href="/singleplayer"
        onClick={[Function]}
      >
        Single Player
      </a>
    </div>
    <div
      className="row justify-content-center border-top py-3"
    >
      <a
        className="btn btn-home"
        href="/multiplayer"
        onClick={[Function]}
      >
        Local Multiplayer
      </a>
    </div>
    <div
      className="row justify-content-center border-top py-3"
    >
      <a
        className="btn btn-home"
        href="/online"
        onClick={[Function]}
      >
        <span>
          Online
        </span>
      </a>
    </div>
    <div
      className="row justify-content-center border-top py-3"
    >
      <a
        className="btn btn-home"
        href="/tutorial"
        onClick={[Function]}
      >
        How to play
      </a>
    </div>
    <div
      className="row justify-content-center border-top py-3"
    >
      <a
        className="btn btn-home"
        href="/contributors"
        onClick={[Function]}
      >
        <span>
          Contributors
        </span>
      </a>
    </div>
  </div>
  <div
    className="container"
  >
    <div
      className="row text-center pb-3 border-bottom"
    >
      <div
        className="col"
      >
        <h2>
          How to play
        </h2>
      </div>
    </div>
    <div
      className="pt-3"
    />
    <p>
      The mechanics of the game are simple you have to do a click in any of the cells, in any of the available grids, that are those which have the white background.
    </p>
    <p>
      The objective of the game is to have three of your symbols together in diagonal, row or column in one of the grids.
    </p>
    <p>
      The first move is always in whatever grid and cell you want, after that the next move is decided based on the previous movement. Look at this example:
    </p>
    <div
      className="row text-center"
    >
      <div
        className="col"
      >
        <img
          alt="example"
          className="img-fluid"
          src="/src/images/example1.gif"
        />
      </div>
    </div>
    <p>
      In this previous example, the player X could make any move that it wants (the first player is always the X), it chooses the middle cell at the bottom, then the player O has to make a move in the middle grid at the bottom.
    </p>
    <div
      className="row text-center"
    >
      <div
        className="col"
      >
        <img
          alt="example"
          className="img-fluid"
          src="/src/images/example2.png"
        />
      </div>
    </div>
    <p>
      In the previous image, we can see that the next move has to be at the left bottom grid because it is the only grid with a white background. And that the previous move was done in the center or top right grid because those are the only ones that have in the bottom left cell an X or an O.
    </p>
    <div
      className="row text-center"
    >
      <div
        className="col"
      >
        <a
          className="btn btn-game btn-lg btn-danger"
          href="/"
          onClick={[Function]}
        >
          Back
        </a>
      </div>
    </div>
  </div>
  <div
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      {
        "bottom": 16,
        "left": 16,
        "pointerEvents": "none",
        "position": "fixed",
        "right": 16,
        "top": 16,
        "zIndex": 9999,
      }
    }
  />
</div>
`;
